### ISSUE TITLE
Proposal: Support packaging local dependencies in crate files

### ISSUE THREAD

**Issue by joshlf** [rust-lang/cargo#4313]
TL;DR: When publishing a crate, instead of automatically also publishing its filesystem-local dependencies, package those dependencies in the crate file.

NOTE: There's some previous discussion on this topic [here](https://users.rust-lang.org/t/publish-crate-without-publishing-local-sub-crates/11299).

---

I have a repository with two top-level crates that I intend to publish. I also have one crate that the first two depend on, but I do not want to publish it. Currently, I have a few options:
* Publish it and commit to maintaining it at a level of quality and API completeness much higher than is needed for what I'm using it for.
* Publish it, don't commit to maintaining it, and be responsible for cluttering up the crates.io namespace.
* Duplicate the code in each of the top-level crates so that I don't need a separate third crate as a dependency.
* Only have one copy of the code, but instead of making it its own crate, use symlinks to make it seem like a submodule of both of the top-level crates. Ew.

All of these options are pretty undesirable.

My proposal is to extend the way we package crates so that filesystem-local dependencies are automatically bundled in the crate file (the crate file is [just a tarball](https://users.rust-lang.org/t/where-to-learn-about-crate-file-format/11496/7?u=joshlf), so this should be easy). Then, when a crate is downloaded, its local dependencies will be right there, and should Just Work (or at least, it should require relatively little work to get the existing tools to support it).

**Comment by epage** [comment]
This looks to be the same as #2203, closing in favor of that.  If there is something I missed, let us know!
